### DISCUSSION TITLE
Can I safely use the 128 byte RIOT memory when coding for the Atari 7800 ?

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#1732] [Q&A] [answered]
@karrika Can I use the 128 byte RIOT memory when coding for the Atari 7800 ? I am scavenging for memory.

I know how to do it. I know map variables to specific memory locations. 
I wonder whether those locations can be used safely and are not used in any way by the Atari 7800 in non-VCS mode.

**Comment by karrika** [comment]
Look at the atari7800.cfg file. The ram is defined there. You cannot omit the zeropage area, the hw stack or the C-stack.

But you can define all your data arrays const so that they do not use ram at all. That is how I was able to squeeze Pirate Cove in the 7800.

**Comment by Fabrizio-Caruso** [comment]
@karrika I don't see the RIOT 128 bytes in the cfg file. These are the (emulated) VCS ram memory. They are usually not used in Atari 7800 mode.
They should be in $280-$2FF. I could add a memory section for them or map some of my variables in those locations with an Assembly file, some pragmas and some export directives. 
My question is whether doing this is safe. I think I know how to do these sort of things.
I do not know the Atari 7800 well enough. I don't know whether the Atari 7800 uses them behind the scenes also in Atari 7800 mode.

**Comment by karrika** [comment]
There is just a single 4k RAM chip in the 7800 machine. It is at the address $1800 and extends to $2800.

1800 to 203F | RAM1
2040 to 20FF | Block Zero Shadow. This is seen as zeropage $40 to $ff (ZP)
2100 to 213F | RAM2
2140 to 21FF | Block One Shadow. This is seen as hardware stack $140 to $1ff (SP)
2200 to 27FF | RAM3

```
    ZP:     file = "", define = yes, start = $0040, size = $00C0, type = rw;
    SP:     file = "", define = yes, start = $0140, size = $00C0, type = rw;
    RAM1:   file = "", define = yes, start = $1800, size = $0840, type = rw;
    RAM2:   file = "", define = yes, start = $2100, size = $0040, type = rw;
    RAM3:   file = "", define = yes, start = $2200, size = $0600, type = rw;

```

Currently my code sets RAM1 to be usable RAM for the programs.
RAM2 is very small. It is currently not used for anything.
RAM3 is used for the C-stack.

You might use RAM2 for the C-stack and use RAM3 for your programs.

The 6532 chip is too slow for the 7800. That is why it is only used for switches and joysticks. The CPU slows down from 1.79MHz to 1.19MHz. So it is not recommended to use it.

If you plan to use it anyway the address range is $480 to $4ff.

**Comment by Fabrizio-Caruso** [comment]
I usually compile with -Cl which reduces the use of the C stack a lot. I think I got it. I know how to reduce the size of the C stack. It is quite big by default.

**Comment by karrika** [comment, marked as answer]
You just need to rename RAM2 to RAM3 in the atari7800.cfg file. The crt0 will set the stack to RAM3.

Currently it is:
```
        ; Set up parameter stack
        lda     #<(__RAM3_START__ + __RAM3_SIZE__)
        sta     sp
        lda     #>(__RAM3_START__ + __RAM3_SIZE__)
        sta     sp+1
```
But the risk here is that some C-programs may run out of stack and heap area if the area is too small.

**Comment by Fabrizio-Caruso** [comment]
I never use the heap in game code written for Cross-Lib. It is forbidden in Cross-Lib as some of the compilers do not support it. I usually compile with -Cl (all locals become static). So the C stack in my usecases can be made much smaller.
